File: config/model_registry.py
from __future__ import annotations


class ModelRegistry:
    """Provides model lookup by id, purpose, cost_tier."""

    def __init__(self, config: dict):
        self._models = config.get("models", [])
        self._routing = config.get("routing", {})

    def get_model_by_id(self, model_id: str) -> dict | None:
        for model in self._models:
            if model["id"] == model_id:
                return model
        return None

    def get_default_model(self) -> dict | None:
        default_id = self._routing.get("default_model")
        if default_id:
            return self.get_model_by_id(default_id)
        for model in self._models:
            if model.get("default"):
                return model
        return None

    def get_fallback_model(self) -> dict | None:
        fallback_id = self._routing.get("fallback_model")
        if fallback_id:
            return self.get_model_by_id(fallback_id)
        return self.get_default_model()

    def get_models_by_purpose(self, purpose: str) -> list[dict]:
        return [m for m in self._models if purpose in m.get("purpose", [])]

File: src/config/model-registry.ts
import type { ModelConfig } from './loader.js'

/** 模型注册表：按 id 索引模型配置，支持默认/fallback 模型和按 purpose 过滤 */
export class ModelRegistry {
  private models: Map<string, ModelConfig> = new Map()
  private defaultModel: ModelConfig | null = null
  private fallbackModel: ModelConfig | null = null

  constructor(models: ModelConfig[], defaultModelId: string, fallbackModelId: string) {
    for (const model of models) {
      this.models.set(model.id, model)
      if (model.id === defaultModelId) this.defaultModel = model
      if (model.id === fallbackModelId) this.fallbackModel = model
    }
  }

  /** 通过 id 查找模型 */
  getModelById(id: string): ModelConfig | undefined {
    return this.models.get(id)
  }

  /** 获取默认模型 */
  getDefaultModel(): ModelConfig | null {
    return this.defaultModel
  }

  /** 获取 fallback 模型 */
  getFallbackModel(): ModelConfig | null {
    return this.fallbackModel
  }

  /** 按 purpose 过滤模型 */
  getModelsByPurpose(purpose: string): ModelConfig[] {
    return [...this.models.values()].filter(m => m.purpose.includes(purpose))
  }
}
